pkg/registry/core/service/storage/storage.go

```diff
@@ -648,9 +654,19 @@ func needsClusterIP(svc *api.Service) bool {
 }
 
 func needsNodePort(svc *api.Service) bool {
-	if svc.Spec.Type == api.ServiceTypeNodePort || svc.Spec.Type == api.ServiceTypeLoadBalancer {
+	if svc.Spec.Type == api.ServiceTypeNodePort {
 		return true
 	}
+	if svc.Spec.Type == api.ServiceTypeLoadBalancer {
+		if svc.Spec.AllocateLoadBalancerNodePorts == nil || *svc.Spec.AllocateLoadBalancerNodePorts {
+			return true
+		}
+		for _, item := range svc.Spec.Ports {
+			if item.NodePort > 0 {
+				return true
+			}
+		}
+	}
```

Comment:
I'm not sure this heuristic here is correct, there are at least other function in strategy that replicates this logic that we need to check too

Reply:
i'll take a look the strategy part, and add corresponding integration tests, e2e tests.

Reply:
to be clear, I don't think that the existence of NodePorts should be part of this function,

Reply:
right. after digging more, this part is not required. removed.